# Meeting Notes from 03.06.2024

Get Intersection Hashes for Protein Family 
- Input: Protein Family (Mapman Bin) ID (BINCODE)
- Output: Hashes that appear in all Proteins of that Mapman Bin
- Algorithm:
  prot_ids = Get protein Ids for Bincode
  prot_id_hashes = [] # Array of math. sets
  for each p_id in prot_ids:
    p_hshs = get hashes for p_id
    append p_hshs to prot_id_hashes
  Reduce(intersect, prot_id_hashes) # pairwise set intersection
  \# or maybe in Python intersect(prot_id_hashes)  
  
  
Eine Möglichkeit für eine effiziente DB:
- SQLite: _mit_ Indizes 
- Tables 
  - hashes: hash
  - protein: id (UniProt), sequence, length
  - prot_to_hash
    - protein_id FK
    - hash_id FK
    
Oder in Pickle:
- Dictionary: protein: [hashes]

BASH Kommandos zur Ermittlung von Mapman Bin
Statistiken:

awk -F "\t" 'BEGIN{f=""; c=0}{if ($5 == "T"){if(f != $1){print f "\t" c; c=1;f=$1}else{c=c+1} }}' mercatorresult.txt | grep -cP '\t1$'

Gene Family Databases:
Two optionsL EGGNOG and InterPro

InterPro:
- https://www.ebi.ac.uk/interpro/download/
- filter entrylist for ENTRY_TYPE 
sed -e '1d' entry.list | awk -F "\t" '{if ($2 == "Family" || $2 == "Homologous_superfamily") {print} }' > interpro_families.txt
- get proteins for above families from this file: https://ftp.ebi.ac.uk/pub/databases/interpro/current_release/protein2ipr.dat.gz

EGGNOG:
- orthogroups from EGGNOG: http://eggnog5.embl.de/download/eggnog_5.0/
  * http://eggnog5.embl.de/download/eggnog_5.0/   FTP site
  * http://eggnog5.embl.de/#/app/guided_search
  
  
One other option
- BUSCO single copy orthologs: run busco on human, mouse, arabidopsis, rice, s. cerevisiae, e. coli 
  or look for reference databases
  
  
REFERENCES:
- Find good references based on Phylogenetic relatedness, i.e. gene families that are comprised of genes that have a last common ancestor
- Scan literature for such reference databases (max four hours)
- Rossier, V., Warwick Vesztrocy, A., Robinson-Rechavi, M., & Dessimoz, C. (2021). OMAmer: Tree-driven and alignment-free protein assignment to subfamilies outperforms closest sequence approaches. Bioinformatics, 37(18), 2866–2873. https://doi.org/10.1093/bioinformatics/btab219
- Emms, D. M., & Kelly, S. (2015). OrthoFinder: Solving fundamental biases in whole genome comparisons dramatically improves orthogroup inference accuracy. Genome Biology, 16, 157. https://doi.org/10.1186/s13059-015-0721-2
- Emms, D. M., & Kelly, S. (2019). OrthoFinder: Phylogenetic orthology inference for comparative genomics. Genome Biology, 20(1), 238. https://doi.org/10.1186/s13059-019-1832-y

R Code
```R
n_hashes <- function(n, f_per_win=3) { n(n+1)/2 * f_per_win }
n_win <- seq(5,50,1)
plot(n_win, sapply(n_win, n_hashes), type="b")
n_hashes <- function(n, f_per_win=3) { n*(n+1)/2 * f_per_win }
hash_size <- ceiling(log(25,2))*2 + 2 + ceiling(log(15,2))
# window dist max 15
n_hashes(15) * hash_size / (8 * 1024^3)
n_entries_uniprot_kb <- 244910918
uniprotkb_prot_fin_db_size <- n_hashes(15) * hash_size * n_entries_uniprot_kb / (8 * 1024^3)
uniprotkb_prot_fin_db_size
```

To Dos:
- Family Approach to be corrected
- Test restriction of DB size: see max window distance (R code)
  - possibly also max no. or proteins per hash: do not do both experiments in one
  - How well do we still find our proteins? Unique self matches (best) and best score for self match (still good)
  - How well do we find intersection hashes with these reduced dbs?
    - If not, this is not that bad, because we can build our DBs for specific use cases
- Once we have that, we can start testing our InterPro, Eggnog, BUSCO families
